meson.build

```diff
@@ -0,0 +1,484 @@
+    'xkbcommon/xkbcommon-names.h',
+    subdir: 'xkbcommon',
+)
+pkgconfig.generate(
+    name: 'xkbcommon',
+    filebase: 'xkbcommon',
+    libraries: libxkbcommon,
+    version: meson.project_version(),
+    description: 'XKB API common to servers and clients',
+)
+
+
+# libxkbcommon-x11.
+if get_option('enable-x11')
+    xcb_dep = dependency('xcb', version: '>=1.10', required: false)
+    xcb_xkb_dep = dependency('xcb-xkb', version: '>=1.10', required: false)
+    if not xcb_dep.found() or not xcb_xkb_dep.found()
+        error('''X11 support requires xcb-xkb >= 1.10 which was not found.
+You can disable X11 support with -Denable-x11=false.''')
+    endif
+
+    libxkbcommon_x11_internal = static_library(
+        'xkbcommon-x11-internal',
+        'src/x11/keymap.c',
+        'src/x11/state.c',
+        'src/x11/util.c',
+        'src/x11/x11-priv.h',
+        'src/context.h',
+        'src/context-priv.c',
+        'src/keymap.h',
+        'src/keymap-priv.c',
+        'src/atom.h',
+        'src/atom.c',
+        include_directories: include_directories('src'),
+        link_with: libxkbcommon,
+        dependencies: [
+            xcb_dep,
+            xcb_xkb_dep,
+        ],
+    )
+    libxkbcommon_x11_link_args = []
+    if have_version_script
+        libxkbcommon_x11_link_args += '-Wl,--version-script=' + join_paths(meson.source_root(), 'xkbcommon-x11.map')
+    endif
+    libxkbcommon_x11 = library(
+        'xkbcommon-x11',
+        'xkbcommon/xkbcommon-x11.h',
+        link_whole: libxkbcommon_x11_internal,
+        link_args: libxkbcommon_x11_link_args,
+        link_depends: 'xkbcommon-x11.map',
+        version: '0.0.0',
+        install: true,
+    )
+    install_headers(
+        'xkbcommon/xkbcommon-x11.h',
+        subdir: 'xkbcommon',
+    )
+    pkgconfig.generate(
+        name: 'xkbcommon-x11',
+        filebase: 'xkbcommon-x11',
+        libraries: libxkbcommon_x11,
+        version: meson.project_version(),
+        description: 'XKB API common to servers and clients - X11 support',
+    )
+endif
+
+
+# Tests
+test_env = environment()
+test_env.set('XKB_LOG_LEVEL', 'debug')
+test_env.set('XKB_LOG_VERBOSITY', '10')
+test_env.set('top_srcdir', meson.source_root())
+test_env.set('MALLOC_PERTURB_', '15')
+test_env.set('MallocPreScribble', '1')
+test_env.set('MallocScribble', '1')
+# Some tests need to use unexported symbols, so we link them against
+# the internal copy of libxkbcommon with all symbols exposed.
+libxkbcommon_test_internal = static_library(
+    'xkbcommon-test-internal',
+    'test/common.c',
+    'test/test.h',
+    'test/evdev-scancodes.h',
+    include_directories: include_directories('src'),
+    link_with: libxkbcommon_internal,
+)
+test_dep = declare_dependency(
+    include_directories: include_directories('src'),
+    link_with: libxkbcommon_test_internal,
+)
+test(
+    'keysym',
+    executable('test-keysym', 'test/keysym.c', dependencies: test_dep),
+    env: test_env,
+)
+test(
```

Comment:
I started out with a for loop over an array of arrays but I didn't like it much. You lose the flexibility to modify a specific test, and it makes things just a little more opaque.

 I think the repetition is not so bad, I'm trained by automake to accept that :)